src/raw/cgroupv1.go

```diff
@@ -38,20 +49,18 @@ func (cg *CgroupsV1Fetcher) Fetch(c types.ContainerJSON) (Metrics, error) {
 	pid := c.State.Pid
 	containerID := c.ID
 
-	var (
-		cgroupInfo *cgroupV1Paths
-		err        error
-	)
-	cgroupInfo, err = getCgroupV1Paths(cg.hostRoot, pid)
-
+	err := cg.cgroupDetector.PopulatePaths(cg.hostRoot, pid)
 	if err != nil {
 		return stats, err
 	}
 
+	cgroupInfo := cg.cgroupDetector.(*CgoupsV1Detector).paths
```

Comment:
As I see it, `CgroupDetector` interface could be a good idea because it would allow us to mock and uninttest fetchers. However, v1 and v2 detectors results are quite different and, as a result, castings are needed. For this reason, I think we do not get benefit from `CgroupDetector` abstraction. Furthermore, If we tried to implement a mock by implementing current `CgroupDetector` we would not be able to use it in fetchers because we are already casting to the specific type. 

Therefore, I would keep references to the specific structs instead, something like this for cgroupv1:

```go
// pathsDetector would have current *CgroupsV1Detector type, we may improve naming
paths, err := cg.pathsDetector.Paths()
```

If we decide to mock this in order to test fetchers, then I would implement the abstraction by defining different interfaces for `*CgroupsV1Detector` and `*CgroupsV2Detector`.

What do you think?

Reply:
You are right.
I have implemented interfaces per CgroupsVersion, both for the detector and the paths, I think now it gets much better.